File: sq-800/parse_env.c
#include <stdio.h>
#include <stdint.h>
#include <string.h>

uint16_t modbusCRC(uint8_t *buffer, int len)
{
  uint16_t crc = 0xFFFF;
  for (int pos = 0; pos < len; pos++)
  {
    crc ^= (uint16_t)buffer[pos];
    for (int i = 8; i > 0; i--)
    { // 使用 > 保证循环次数明确
      if ((crc & 0x0001) != 0)
      {
        crc >>= 1;
        crc ^= 0xA001;
      }
      else
      {
        crc >>= 1;
      }
    }
  }
  return crc;
}

struct sensor_data
{
  float dsp310_temperature;
  float pressure;
  uint16_t co2;
  uint16_t pm2p5;
  uint16_t pm10;
  int16_t temperature;
  uint16_t humidity;
};

struct louver_data_struct_v2
{
  uint8_t deviceId[6];       // 6 bytes
  uint8_t reserved[10];      // 10 bytes (corrected comment)
  uint16_t pm2p5;            // 2 bytes
  uint16_t pm10;             // 2 bytes
  uint16_t co2;              // 2 bytes
  uint16_t humidity;         // 2 bytes
  float hdc1080_temperature; // 4 bytes
  float dsp310_temperature;  // 4 bytes
  float airPressure;         // 4 bytes
};

typedef struct sensor_data sensor_data_t;
typedef struct louver_data_struct_v2 louver_data_struct_v2_t;

sensor_data_t global_sensor_result;
louver_data_struct_v2_t global_louver_result;

int main()
{
  // Response buffer (fill with your actual response data)
  uint8_t response[] = {0x10, 0x20, 0x15, 0xFC, 0xF2, 0xFB, 0x41, 0xBD, 0x6B, 0xC5, 0x47, 0x88, 0x0B, 0x02, 0x00, 0x0A, 0x00, 0xF9, 0x00, 0xD6, 0x02, 0x00, 0x00, 0x01, 0x61, 0xF5};
  uint8_t response_v2[] = {0x01, 0x21, 0x25, 0x50, 0xE3, 0xB6, 0x6E, 0x9E, 0x9C, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x01, 0x00, 0xA1, 0x07, 0x9B, 0x02, 0x1C, 0xC6, 0x90, 0x43, 0xB8, 0xD6, 0xEB, 0x41, 0xBE, 0xDA, 0xC4, 0x47, 0x02, 0xFF, 0x47}; // Interpret the response

  if (modbusCRC(response, sizeof(response)) == 0)
  {
    memcpy(&global_sensor_result, &response[3], sizeof(global_sensor_result));
    printf("PM2.5=%d ug/m3\n", global_sensor_result.pm2p5);
    printf("PM10=%d ug/m3\n", global_sensor_result.pm10);
    printf("CO2=%d ppm\n", global_sensor_result.co2);
    printf("Humidity=%d %%\n", global_sensor_result.humidity);
    printf("Temperature=%.2f C\n", global_sensor_result.temperature);
    printf("Air Pressure=%.2f hPa\n", global_sensor_result.pressure);
  }
  else
  {
    printf("CRC error\n");
  }

  if (modbusCRC(response_v2, sizeof(response_v2)) == 0)
  {
    memcpy(&global_louver_result, &response_v2[3], sizeof(global_louver_result));
    printf("PM2.5=%d ug/m3\n", global_louver_result.pm2p5);
    printf("PM10=%d ug/m3\n", global_louver_result.pm10);
    printf("CO2=%d ppm\n", global_louver_result.co2);
    printf("Humidity=%d %%\n", global_louver_result.humidity);
    printf("HDC1080 Temperature=%.2f C\n", global_louver_result.hdc1080_temperature);
    printf("DSP310 Temperature=%.2f C\n", global_louver_result.dsp310_temperature);
    printf("Air Pressure=%.2f hPa\n", global_louver_result.airPressure);
  }
  else
  {
    printf("CRC error\n");
  }

  return 0;
}

File: sq-800/parse_env.py
import struct
louver_format = '<2f3HhH'
louver_format_v2 = '<6s10s4H3f'

def modbus_crc(buffer):
    crc = 0xFFFF
    for pos in buffer:
        crc ^= pos
        for _ in range(8):
            if crc & 0x0001:
                crc >>= 1
                crc ^= 0xA001
            else:
                crc >>= 1
    return crc

class SensorData:
    def __init__(self, data):
        self.temperature, self.pressure, self.co2, self.pm2p5, self.pm10, self.temperature_adc, self.humidity = struct.unpack(louver_format, data)

class LouverDataStructV2:
    def __init__(self, data):
        self.deviceId, self.reserved, self.pm2p5, self.pm10, self.co2, self.humidity, self.hdc1080_temperature, self.dsp310_temperature, self.airPressure = struct.unpack(louver_format_v2, data)
def main():
    response = bytearray([0x10, 0x20, 0x15, 0xFC, 0xF2, 0xFB, 0x41, 0xBD, 0x6B, 0xC5, 0x47, 0x88, 0x0B, 0x02, 0x00, 0x0A, 0x00, 0xF9, 0x00, 0xD6, 0x02, 0x00, 0x00, 0x01, 0x61, 0xF5])
    response_v2 = bytearray([0x01, 0x21, 0x25, 0x50, 0xE3, 0xB6, 0x6E, 0x9E, 0x9C, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x01, 0x00, 0xA1, 0x07, 0x9B, 0x02, 0x1C, 0xC6, 0x90, 0x43, 0xB8, 0xD6, 0xEB, 0x41, 0xBE, 0xDA, 0xC4, 0x47, 0x02, 0xFF, 0x47])

    if modbus_crc(response) == 0:
        data=response[3:(len(response)-5)]
        print(data)
        print(len(data))
        if len(data) == struct.calcsize(louver_format):
            global_sensor_result = SensorData(data)
            print(f"PM2.5={global_sensor_result.pm2p5} ug/m3")
            print(f"PM10={global_sensor_result.pm10} ug/m3")
            print(f"CO2={global_sensor_result.co2} ppm")
            print(f"Humidity={global_sensor_result.humidity} %")
            print(f"Temperature={global_sensor_result.temperature:.2f} C")
            print(f"Air Pressure={global_sensor_result.pressure:.2f} hPa")
        else:
            print("len error")
            print(len(data))
            print(struct.calcsize(louver_format))
    else:
        print("CRC error")

    if modbus_crc(response_v2) == 0:
        data=response_v2[3:(len(response_v2)-3)]
        print(data)
        if len(data) == struct.calcsize(louver_format_v2):
            global_louver_result = LouverDataStructV2(data)
            print(f"PM2.5={global_louver_result.pm2p5} ug/m3")
            print(f"PM10={global_louver_result.pm10} ug/m3")
            print(f"CO2={global_louver_result.co2} ppm")
            print(f"Humidity={global_louver_result.humidity} %")
            print(f"HDC1080 Temperature={global_louver_result.hdc1080_temperature:.2f} C")
            print(f"DSP310 Temperature={global_louver_result.dsp310_temperature:.2f} C")
            print(f"Air Pressure={global_louver_result.airPressure:.2f} hPa")
        else:
            print("len error")
            print(len(data))
            print(struct.calcsize(louver_format_v2))

    else:
        print("CRC error")

if __name__ == "__main__":
    main()